src/cryptonote_core/service_node_quorum_cop.cpp

```diff
@@ -41,165 +41,352 @@
+        {
+          LOG_PRINT_L1("Vote does not reference valid block hash: " << vote.checkpoint.block_hash);
+          return false;
+        }
+      }
+      break;
+    }
 
-    const std::shared_ptr<const quorum_checkpointing> state = m_core.get_checkpointing_quorum(height);
-    if (!state)
+    // NOTE: Only do validation that relies on access cryptonote::core here in quorum cop, the rest goes in voting pool
+    std::shared_ptr<const testing_quorum> quorum = m_core.get_testing_quorum(vote.type, vote.block_height);
+    if (!quorum)
     {
       // TODO(loki): Fatal error
-      LOG_ERROR("Quorum state for height: " << height << " was not cached in daemon!");
-      return;
+      LOG_ERROR("Quorum state for height: " << vote.block_height << " was not cached in daemon!");
+      return false;
     }
 
-    auto it = std::find(state->quorum_nodes.begin(), state->quorum_nodes.end(), my_pubkey);
-    if (it == state->quorum_nodes.end())
-      return;
+    uint64_t latest_height             = std::max(m_core.get_current_blockchain_height(), m_core.get_target_blockchain_height());
+    std::vector<pool_vote_entry> votes = m_vote_pool.add_pool_vote_if_unique(latest_height, vote, vvc, *quorum);
+    bool result                        = !vvc.m_verification_failed;
 
-    //
-    // NOTE: I am in the quorum, handle checkpointing
-    //
-    size_t my_index_in_quorum           = it - state->quorum_nodes.begin();
-    service_nodes::checkpoint_vote vote = {};
-    if (!cryptonote::get_block_hash(block, vote.block_hash))
+    if (!vvc.m_added_to_pool) // NOTE: Not unique vote
+      return result;
+
+    switch(vote.type)
     {
-      // TODO(loki): Fatal error
-      LOG_ERROR("Could not get block hash for block on height: " << height);
-      return;
-    }
+      default:
+      {
+        LOG_PRINT_L1("Unhandled vote type with value: " << (int)vote.type);
+        assert("Unhandled vote type" == 0);
+        return false;
+      };
 
-    vote.block_height        = height;
-    vote.voters_quorum_index = my_index_in_quorum;
-    crypto::generate_signature(vote.block_hash, my_pubkey, my_seckey, vote.signature);
+      case quorum_type::uptime_deregister:
+      {
+        if (votes.size() >= UPTIME_MIN_VOTES_TO_KICK_SERVICE_NODE)
+        {
+          cryptonote::tx_extra_service_node_deregister deregister;
+          deregister.block_height       = vote.block_height;
+          deregister.service_node_index = vote.deregister.worker_index;
+          deregister.votes.reserve(votes.size());
+
+          for (pool_vote_entry const &pool_vote : votes)
+          {
+            cryptonote::tx_extra_service_node_deregister::vote tx_vote = {};
+            tx_vote.validator_index                                    = pool_vote.vote.index_in_group;
+            tx_vote.signature                                          = pool_vote.vote.signature;
+            deregister.votes.push_back(tx_vote);
```

Comment:
add a constructor to `tx_extra_service_node_deregister::vote`, and use:

```c++
std::transform(
    votes.begin(),
    votes.end(),
    std::back_inserter(deregister.votes),
    [](const auto& pool_vote) { 
          return cryptonote::tx_extra_service_node_deregister::vote(
               pool_vote.vote.signature, 
               pool_vote.vote.index_in_group); 
     }
);

Reply:
Added